### PR TITLE
C# SDK tests

### PR DESCRIPTION
**Pull request by RReverser**
# Description of Changes

Ports SDK tests to C#.

In the process had to modify the test runner to cache compilation results and not just binding generation results - this was necessary because .NET doesn't expect same project to be built in parallel via same commands, but as a side benefit speeds up both Rust and C# tests as it doesn't even have to try to build same project again and instead reuses prebuilt Wasm for all other stages.

Note that this also modifies the `publish` command so that if project path is a file that ends in `.wasm` then it uses it as-is instead of building it as a dir containing a project. (cc @cloutiertyler can you confirm if it's okay?)

# API and ABI breaking changes

If this is an API or ABI breaking change, please apply the
corresponding GitHub label.

# Expected complexity level and risk

2

*How complicated do you think these changes are? Grade on a scale from 1 to 5,
where 1 is a trivial change, and 5 is a deep-reaching and complex change.*

*This complexity rating applies not only to the complexity apparent in the diff,
but also to its interactions with existing and future code.*

*If you answered more than a 2, explain what is complex about the PR,
and what other components it interacts with in potentially concerning ways.*

### CODE REVIEW

File: crates/cli/src/subcommands/publish.rs
Comment (gefjon) [inline crates/cli/src/subcommands/publish.rs:144]: Could you extend the `.help` message (or add a `.long_help`) to the `-p` flag which describes this behavior?

File: crates/testing/src/sdk.rs
Comment (gefjon) [inline crates/testing/src/sdk.rs:41]: Do we not print it when a test fails?

**Review by gefjon — APPROVED** [review]
Perhaps you should hold off merging until @cloutiertyler weighs in on the new extension to `spacetime publish`, but it seems reasonable to me. The tests look good, and I like how you've memoized client compilation. Thanks for doing this; I'm glad we're getting tests in to verify that C# and Rust modules are interchangeable.

File: crates/testing/src/sdk.rs
Comment (RReverser) [reply crates/testing/src/sdk.rs:41]: No, Cargo uses test names from corresponding functions. Before this PR this field was used in compilation messages, but now that they're memoized and shared between tests, it seemed misleading to use arbitrary test name (whichever happens first) in that context, so I changed them to just messages like "(compiling)".

Now this field is completely unused, but it's assigned in a bunch of spaces and I wasn't sure whether it's worth removing it altogether from all those places or whether we want to use it again in the future.

**Comment by cloutiertyler** [comment]
> Note that this also modifies the `publish` command so that if project path is a file that ends in `.wasm` then it uses it as-is instead of building it as a dir containing a project. (cc @cloutiertyler can you confirm if it's okay?)

I don't see a problem with this per se, but I also don't see why we couldn't just have a different parameter for --wasm-module instead of --project-path.

File: crates/cli/src/subcommands/publish.rs
Comment (cloutiertyler) [reply crates/cli/src/subcommands/publish.rs:144]: See my comment above.

**Comment by RReverser** [comment]
> I don't see a problem with this per se, but I also don't see why we couldn't just have a different parameter for --wasm-module instead of --project-path.

We could, but mainly not sure of the value of having separate parameters, given that they're mutually exclusive and mean more or less the same thing (path to the project). We already auto-detect language of the project (Rust vs C#), so detecting precompiled Wasm seems in line with that as just another language.

**Comment by cloutiertyler** [comment]
> > I don't see a problem with this per se, but I also don't see why we couldn't just have a different parameter for --wasm-module instead of --project-path.
> 
> We could, but mainly not sure of the value of having separate parameters, given that they're mutually exclusive and mean more or less the same thing (path to the project). We already auto-detect language of the project (Rust vs C#), so detecting precompiled Wasm seems in line with that as just another language.

In that case, I would leave it to your discretion. Keep in mind that it's a decision we'll have to live with for a long time.

**Pull request merged**
